File: mac-app/Tests/QuartzTeachersTests/SectionNumbersValidationTests.swift
import XCTest
@testable import QuartzTeachers

/// Saying what is wrong with the timetable sections, as typed.
final class SectionNumbersValidationTests: XCTestCase {

    // MARK: - Functions

    @MainActor
    func testGoodInputPassesQuietly() {
        XCTAssertNil(NewCourseWizardView.sectionNumbersProblem("1"))
        XCTAssertNil(NewCourseWizardView.sectionNumbersProblem("1,3"))
        XCTAssertNil(NewCourseWizardView.sectionNumbersProblem(" 1 , 3 , 5 "), "Spaces around commas are fine")
    }

    @MainActor
    func testSpacesInsteadOfCommasGetTheCommaHint() {
        let problem: String? = NewCourseWizardView.sectionNumbersProblem("1 3 5")
        XCTAssertEqual(problem, "Use commas between section numbers — e.g. 1,3,5.")
    }

    @MainActor
    func testAStrayLetterIsNamed() {
        let problem: String? = NewCourseWizardView.sectionNumbersProblem("1a,3 4")
        XCTAssertNotNil(problem)
        XCTAssertTrue(problem?.contains("“1a”") == true, "The bad piece is quoted so it can be found: \(problem ?? "")")
    }

    @MainActor
    func testTheSilentDropCaseIsCaught() {
        // "1,3 5" used to parse as just [1], quietly losing the rest.
        XCTAssertNotNil(NewCourseWizardView.sectionNumbersProblem("1,3 5"))
    }

    @MainActor
    func testEmptyAndPunctuationShapes() {
        XCTAssertNotNil(NewCourseWizardView.sectionNumbersProblem(""))
        XCTAssertEqual(NewCourseWizardView.sectionNumbersProblem("1,,3"), "There’s an empty spot between commas.")
        XCTAssertEqual(NewCourseWizardView.sectionNumbersProblem("1,"), "There’s an empty spot between commas.")
    }

    @MainActor
    func testZeroAndDuplicates() {
        XCTAssertEqual(NewCourseWizardView.sectionNumbersProblem("0,1"), "“0” isn’t a section number — sections are 1 or higher.")
        XCTAssertEqual(NewCourseWizardView.sectionNumbersProblem("1,3,1"), "Section 1 is listed more than once.")
    }

    @MainActor
    func testATeacherNeedNotTeachSectionOne() {
        // A school may run five sections while this teacher has only three
        // of them — the list does not have to include 1, or be contiguous.
        XCTAssertNil(NewCourseWizardView.sectionNumbersProblem("2,4,5"))
        XCTAssertNil(NewCourseWizardView.sectionNumbersProblem("3"))
        XCTAssertNil(NewCourseWizardView.sectionNumbersProblem("7,2"))
    }

    // MARK: - The code field

    // The course-code cases used to live here, as a copy of the wizard's own
    // rule. They now live in `contracts/course-management.json` under
    // `courseCode`, run by `CourseManagementContractTests`, because RENAMING
    // a course asks the identical question — and a rule written down twice
    // is a rule that will disagree with itself. The wizard consults
    // `CourseCodeRule`; so does the sidebar's rename field.

}

File: windows-app/Plantoir.Tests/SectionNumbersValidationTests.cs
using Plantoir.Core.Models;
using Xunit;

namespace Plantoir.Tests;

public class SectionNumbersValidationTests
{
    [Fact]
    public void GoodInputPassesQuietly()
    {
        Assert.Null(SectionNumbersRule.Problem("1"));
        Assert.Null(SectionNumbersRule.Problem("1,3"));
        Assert.Null(SectionNumbersRule.Problem(" 1 , 3 , 5 "));
    }

    [Fact]
    public void SpacesInsteadOfCommasGetTheCommaHint()
    {
        string? problem = SectionNumbersRule.Problem("1 3 5");
        Assert.Equal("Use commas between section numbers — e.g. 1,3,5.", problem);
    }

    [Fact]
    public void StrayLetterIsNamed()
    {
        string? problem = SectionNumbersRule.Problem("1a,3 4");
        Assert.NotNull(problem);
        Assert.Contains("“1a”", problem);
    }

    [Fact]
    public void TheSilentDropCaseIsCaught()
    {
        Assert.NotNull(SectionNumbersRule.Problem("1,3 5"));
    }

    [Fact]
    public void EmptyAndPunctuationShapes()
    {
        Assert.NotNull(SectionNumbersRule.Problem(""));
        Assert.Equal("There’s an empty spot between commas.", SectionNumbersRule.Problem("1,,3"));
        Assert.Equal("There’s an empty spot between commas.", SectionNumbersRule.Problem("1,"));
    }

    [Fact]
    public void ZeroAndDuplicates()
    {
        Assert.Equal("“0” isn’t a section number — sections are 1 or higher.", SectionNumbersRule.Problem("0,1"));
        Assert.Equal("Section 1 is listed more than once.", SectionNumbersRule.Problem("1,3,1"));
    }

    [Fact]
    public void TeacherNeedNotTeachSectionOne()
    {
        Assert.Null(SectionNumbersRule.Problem("2,4,5"));
        Assert.Null(SectionNumbersRule.Problem("3"));
        Assert.Null(SectionNumbersRule.Problem("7,2"));
    }
}
